Remove TakeFirstAssetTrader from AH Westend and Rococo

x3c41a opened the pull request:
## Description

Removed `TakeFirstAssetTrader` from Asset Hub Westend and Rococo. Improved macros, fixed tests.

This implies asset sufficiency no longer guarantees that weight can also be bought with it. `SwapFirstAssetTrader` is used instead which will attempt to swap some of the given asset for the required amount of native asset to buy weight. This may or may not succeed depending on whether there is a local pool present with enough liquidity to serve the swap.

## Review notes

Additionally parametrised macro and fixed Westend test: weight swapping was failing at [this line](https://github.com/paritytech/polkadot-sdk/blob/44ae6a8bebd23a8ffac02d71c6e74ee889c3ab00/substrate/frame/asset-conversion/src/lib.rs#L903) with around 100x difference thus had to modify the macro.

Fixes #8233

x3c41a commented on cumulus/parachains/integration-tests/emulated/tests/assets/asset-hub-westend/src/tests/transact.rs:
```diff
@@ -113,7 +115,14 @@ fn transact_from_para_to_para_through_asset_hub() {
 
 	// We create a pool between WND and USDT in AssetHub.
 	let usdt = Location::new(0, [PalletInstance(ASSETS_PALLET_ID), GeneralIndex(USDT_ID.into())]);
-	create_pool_with_wnd_on!(AssetHubWestend, usdt, false, AssetHubWestendSender::get());
```
[The line](https://github.com/paritytech/polkadot-sdk/blob/44ae6a8bebd23a8ffac02d71c6e74ee889c3ab00/substrate/frame/asset-conversion/src/lib.rs#L903) mentioned in the description:
- 3423620724268 - `amount_in` 
- 90000000000 - `amoamount_in_max`

franciscoaguirre replied:
Hmm. Do you know why this happens?
Allowing custom amounts on the macro is a good addition anyway.

x3c41a replied:
My understanding is that the default 1:2 (DOT:USDT) proportion increased the `amount_in` which was already high  because of Transact call weight.

x3c41a commented:
if it comes to `[Check publish build / check-publish-compile (pull_request)](https://github.com/paritytech/polkadot-sdk/actions/runs/14749062513/job/41402146630?pr=8376)` failure, I ran `cargo build -p parachain-template-node` on my local machine and everything went well; might be unrelated to my change.

franciscoaguirre commented on cumulus/parachains/integration-tests/emulated/tests/assets/asset-hub-westend/src/tests/mod.rs:
```diff
@@ -42,8 +42,8 @@ macro_rules! foreign_balance_on {
 #[macro_export]
-macro_rules! create_pool_with_wnd_on {
-	( $chain:ident, $asset_id:expr, $is_foreign:expr, $asset_owner:expr ) => {
+macro_rules! create_pool_with_wnd_on_custom_amounts {
```
You can actually include these new parameters in the same macro, since macros can take any number of arguments.
You can take in optional parameters and use them in this same branch or create a new branch.

x3c41a replied:
Extracted into a separate PR to experiment with macros a bit - https://github.com/paritytech/polkadot-sdk/pull/8392

x3c41a commented:
I modified [runtime test](https://github.com/paritytech/polkadot-sdk/pull/8376/files#diff-c3fa8407df8061aaa679802d6cdd75c10f69762acec2159d58d6a78cec43fcf3) - mainly removed the tests which were testing `take_first_trader` functionality (since it's not needed now).
**Let me know in case I was wrong and we do need to keep those tests**

Also, I fixed `test_asset_xcm_take_first_trader_not_possible_for_non_sufficient_assets` by adding liquidity pool to it.

franciscoaguirre reviewed (approved):
It's fine to remove those tests. We no longer use sufficient assets for fee payment, we always use liquidity pools, and we have tests for using liquidity pools to swap for fee payment already

acatangiu commented on cumulus/parachains/runtimes/assets/asset-hub-westend/src/xcm_config.rs:
```diff
@@ -426,32 +426,6 @@ impl xcm_executor::Config for XcmConfig {
-		// `pallet_assets` instance - `Assets`.
-		cumulus_primitives_utility::TakeFirstAssetTrader<
-			AccountId,
-			AssetFeeAsExistentialDepositMultiplierFeeCharger,
```
Some of these helper types are not used since you remove this trader, please remove them too.
e.g. `AssetFeeAsExistentialDepositMultiplierFeeCharger`, `XcmAssetFeesReceiver`, `ForeignAssetFeeAsExistentialDepositMultiplierFeeCharger`

acatangiu commented on cumulus/parachains/integration-tests/emulated/tests/assets/asset-hub-rococo/src/tests/send.rs:
```diff
@@ -106,79 +106,79 @@ fn send_xcm_from_para_to_asset_hub_paying_fee_with_system_asset() {
-		);
-	});
-}
+// /// We tests two things here:
+// /// - Parachain should be able to send XCM paying its fee at Asset Hub using sufficient asset
+// /// - Parachain should be able to create a new Asset at Asset Hub
+// #[test]
+// fn send_xcm_from_para_to_asset_hub_paying_fee_with_sufficient_asset() {
+// 	let para_sovereign_account = AssetHubRococo::sovereign_account_id_of(
+// 		AssetHubRococo::sibling_location_of(PenpalA::para_id()),
+// 	);
+//
+// 	// Force create and mint sufficient assets for Parachain's sovereign account
+// 	AssetHubRococo::force_create_and_mint_asset(
```
leftover commented code

x3c41a replied:
tests are fixed

acatangiu commented on cumulus/parachains/integration-tests/emulated/tests/assets/asset-hub-westend/src/tests/send.rs:
```diff
@@ -106,79 +106,90 @@ fn send_xcm_from_para_to_asset_hub_paying_fee_with_system_asset() {
-		);
-	});
-}
+// /// We tests two things here:
+// /// - Parachain should be able to send XCM paying its fee at Asset Hub using sufficient asset
+// /// - Parachain should be able to create a new Asset at Asset Hub
+// #[test]
+// fn send_xcm_from_para_to_asset_hub_paying_fee_with_sufficient_asset() {
+// 	let para_sovereign_account = AssetHubWestend::sovereign_account_id_of(
+// 		AssetHubWestend::sibling_location_of(PenpalA::para_id()),
+// 	);
+//
+// 	// Force create and mint sufficient assets for Parachain's sovereign account
+// 	AssetHubWestend::force_create_and_mint_asset(
```
same here

x3c41a replied:
this one too

acatangiu commented on prdoc/pr_8376.prdoc:
```diff
@@ -0,0 +1,20 @@
+title: 'Stopped using TakeFirstAssetTrader in AH Westend and Rococ'
```
```suggestion
title: 'Remove TakeFirstAssetTrader from AH Westend and Rococo'
```

acatangiu commented on prdoc/pr_8376.prdoc:
```diff
@@ -0,0 +1,20 @@
+title: 'Stopped using TakeFirstAssetTrader in AH Westend and Rococ'
+doc:
+- audience: Runtime Dev
+  description: |-
+    Removed `TakeFirstAssetTrader` from configs. Improved macros, fixed tests.
```
```suggestion
- audience: [Runtime Dev, Runtime User]
  description: |-
    Removed `TakeFirstAssetTrader` from Asset Hub Westend and Rococo. Improved macros, fixed tests.
    This implies asset sufficiency no longer guarantees that weight can also be bought with it.
    `SwapFirstAssetTrader` is used instead which will attempt to swap some of the given asset for the
    required amount of native asset to buy weight. This may or may not succeed depending on whether
    there is a local pool present with enough liquidity to serve the swap.
```

x3c41a replied:
> This implies asset sufficiency no longer guarantees that weight can also be bought with it.

What is `asset sufficiency` in this context?
